File: sdk/javascript/src/reliability/health.ts
/**
 * HealthMonitor — formats the exact `client.health()` shape from the
 * ticket:
 *
 *     {
 *       "worker": "running",
 *       "queue_depth": 24,
 *       "retry_queue": 3,
 *       "circuit": "closed",
 *       "compression": "enabled"
 *     }
 */

import type { BackgroundWorker } from "./worker.js";

export interface HealthSnapshot {
  worker: string;
  queue_depth: number;
  retry_queue: number;
  circuit: string;
  compression: string;
}

export interface QueueStatsSnapshot {
  queueDepth: number;
  droppedEvents: number;
  retryQueueSize: number;
  workerStatus: string;
  sentTotal: number;
  failedTotal: number;
  retryCount: number;
  avgUploadLatencyMs: number;
  compressionRatio: number | null;
  lastBatchSize: number;
  workerUptimeSeconds: number;
}

export class HealthMonitor {
  constructor(private readonly worker: BackgroundWorker) {}

  snapshot(): HealthSnapshot {
    const w = this.worker;
    return {
      worker: w.backpressure.workerStatus,
      queue_depth: w.memoryQueue.size,
      retry_queue: w.persistentQueue.count(),
      circuit: w.circuitBreaker.state,
      compression: w.compressionEnabled ? "enabled" : "disabled",
    };
  }

  queueStats(): QueueStatsSnapshot {
    const w = this.worker;
    const metrics = w.metrics.snapshot();
    return {
      queueDepth: w.memoryQueue.size,
      droppedEvents: w.memoryQueue.dropped,
      retryQueueSize: w.persistentQueue.count(),
      workerStatus: w.backpressure.workerStatus,
      ...metrics,
    };
  }
}

File: sdk/python/costorah/reliability/health.py
"""
HealthMonitor — formats the exact `client.health()` shape from the
ticket:

    {
      "worker": "running",
      "queue_depth": 24,
      "retry_queue": 3,
      "circuit": "closed",
      "compression": "enabled"
    }
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from costorah.reliability.worker import BackgroundWorker


class HealthMonitor:
    def __init__(self, worker: BackgroundWorker) -> None:
        self._worker = worker

    def snapshot(self) -> dict[str, Any]:
        w = self._worker
        return {
            "worker": w.backpressure.worker_status,
            "queue_depth": w.memory_queue.qsize(),
            "retry_queue": w.persistent_queue.count() if w.persistent_queue else 0,
            "circuit": w.circuit_breaker.state,
            "compression": "enabled" if w.compression_enabled else "disabled",
        }

    def queue_stats(self) -> dict[str, Any]:
        w = self._worker
        return {
            "queue_depth": w.memory_queue.qsize(),
            "dropped_events": w.memory_queue.dropped_count,
            "retry_queue_size": w.persistent_queue.count() if w.persistent_queue else 0,
            "worker_status": w.backpressure.worker_status,
            **w.metrics.snapshot(),
        }
